### ISSUE TITLE
OTI: Interface functions have no usable type

### ISSUE THREAD

**Issue by SLaks** [google/closure-compiler#1338]
Source:

```js
/** @interface */
function X() {}

/** @param {function(this: X)} a */
function init(a) {}

init(X);
```

Error:

```js
input0:7: WARNING - actual parameter 1 of init does not match formal parameter
found   : function (this:X): ?
required: function (this:X): ?
init(X);
     ^
```

I can't find any type other than `?` that `X` is convertible to (not even `Function`).

Under NTI, it works fine, as `function(new: X)`, as I would expect 

http://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%2540interface%2520*%252F%250Afunction%2520X()%2520%257B%257D%250A%250A%252F**%2520%2540param%2520%257Bfunction(this%253A%2520X)%257D%2520a%2520*%252F%250Afunction%2520init(a)%2520%257B%257D%250A%250Ainit(X)%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26MISSING_PROPERTIES%3D1

**Comment by ChadKillingsworth** [comment]
`X` is the namespace. It would need to be an instance of `X` - > `new X()`; But that's really non-sensical for an interface.

**Comment by SLaks** [comment]
No; `X` is a type, not a namespace.  And this works fine with the new type inference.

**Comment by ChadKillingsworth** [comment]
`init(x)` <-- that's the namespace reference

**Comment by SLaks** [comment]
Huh? `init` is a function, and `X` is a type.

**Comment by ChadKillingsworth** [comment]
Yeah this is complicated.

`X` is a type. It refers to instances of the `function X() {}` method. With a constructor, those would typically by `new X()`. So `new X()` has the type `X`;

'X` also is a symbol (or namespace) that is a static reference. This makes more sense when you add properties:

```js
/** @interface */
function X() {}
X.foo = 1; //X is being referred to as a namespace here.
```

So the tricky part is that a plain old `X` can be both a function or a namespace - but it is not the type `X`.

This is easy to test:

```js
console.log(X instanceof X); //this will of course log false
```

Back to your original issue: `init(X)`. Your init function expects something of the type `X`. I would expect `init(new X())` to work - but since `X` is an interface that's not really expected.

It appears that either OTI cannot treat X as a function and a type reference simultaneously, OR perhaps NTI isn't smart enough to warn for usages like this - and should be.

**Comment by SLaks** [comment]
Read my code more carefully.
`init` expects a "type" (to the extent that JSCompiler supports types as entities), not an instance.

It takes `function(this: X)`, (although `function(new: X)` would be more correct, as NTI does) not `X`.

**Comment by ChadKillingsworth** [comment]
Ahh you are correct - I missed the function() part of the param.

It appears to be a bug specific to interfaces - which are not intended to be directly instantiated. If you have a constructor that implements X and pass that does it work?

I'm thinking perhaps this is NTI bug.
